### PR TITLE
Popups on escape now just close themselves

### PR DESCRIPTION
**Pull request by Raenore**
https://github.com/user-attachments/assets/b61ea850-fc15-4896-87e1-d82580603864

This fixes the fact that escaping a popup would close the whole TRP frame with it as well. Now it behaves similarly to the close button on the popups itself.

Also the popups code is a freaking nightmare, do not recommend.

### CODE REVIEW

**Review by Meorawr — COMMENTED** [review]
I'll look at this tomorrow but one thing we need to consider is that some (all?) popups support cancellation callbacks when the close button is pressed. An escape press should count as a cancel operation and slowly invoke those callbacks if set.

**Comment by Raenore** [comment]
> I'll look at this tomorrow but one thing we need to consider is that some (all?) popups support cancellation callbacks when the close button is pressed. An escape press should count as a cancel operation and slowly invoke those callbacks if set.

I'll cancel you.

But I can look into this, I guess.

**Comment by Raenore** [comment]
I have checked most of these cases, and in just about every situation we currently seem to be feeding `nil` to the browsers for cancel callbacks.

However, I think I've at least figured out how to handle the callbacks specifically for Companion and Pet browsers in the latest commit (I could be wrong ofc, this is all very weird code).
For the others, I don't currently know how to do it properly, so I'll wait for your input/suggestions.

File: totalRP3/Core/Popup.lua [hunk 1]
```diff
@@ -766,8 +766,25 @@ function TRP3_API.popup.showCompanionBrowser(onSelectCallback, onCancelCallback,
...
+		frame:SetScript("OnKeyDown", function(_, key)
+			if key == "ESCAPE" then
+				PlaySound(SOUNDKIT.IG_MAINMENU_OPTION_CHECKBOX_ON);
+				frame:SetPropagateKeyboardInput(false);
+				-- Hiding frames & Cancel callback is handled within.
+				onCompanionClose();
+			else
+				frame:SetPropagateKeyboardInput(true);
+			end
+		end);
```
Comment (Meorawr) [inline totalRP3/Core/Popup.lua:785]: The InCombatLockdown check needs to be inside the SetScript callback, not outside.

**Review by Meorawr — CHANGES_REQUESTED** [review]

File: totalRP3/Core/Popup.lua (hunk 1 above)
Comment (Raenore) [reply totalRP3/Core/Popup.lua]: Should be resolved in latest commit, thank you.

**Comment by Raenore** [comment]
> If cancel callbacks are used nowhere I'd be tempted to just remove support for them entirely. Alternatively, the method of closing each popup could be changed to just `:Click()` each close button - but that feels disgusting.

I mean, if they all are nil/not handled right now then I feel it might just be better to just leave it as-is until the rework?

If the new escape options don't do stuff, and the normal callbacks don't either, there will be a zero sum change at the end. /shrug

**Comment by Meorawr** [comment]
Only issue with leaving them as-is is the risk of us forgetting that they don't consistently work - but similarly, I suspect it unlikely that we'll run into that trap anyway so I'm probably fine with leaving things as-is.

**Review by Solanya — APPROVED** [review]

**Comment by Raenore** [comment]
Allow me to block my own PR for a second to see if I can implement escape closing on the glance editor too.

**Comment by Solanya** [comment]
Sure

**Comment by Raenore** [comment]
Done.

**Review by Meorawr — APPROVED** [review]

**Review by Solanya — APPROVED** [review]

**Pull request merged**